subsys/net/lib/lwm2m/ipso_light_control.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Source material for IPSO Light Control object (3311):
+ * https://github.com/IPSO-Alliance/pub/blob/master/docs/IPSO-Smart-Objects.pdf
+ * Section: "16. IPSO Object: Light Control"
+ */
+
+#define SYS_LOG_DOMAIN "ipso_light_control"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_LWM2M_LEVEL
+#include <logging/sys_log.h>
+#include <stdint.h>
+#include <init.h>
+#include <net/lwm2m.h>
+
+#include "lwm2m_object.h"
+#include "lwm2m_engine.h"
+
+/* Server resource IDs */
+#define LIGHT_ON_OFF_ID				5850
+#define LIGHT_DIMMER_ID				5851
+#define LIGHT_ON_TIME_ID			5852
+#define LIGHT_CUMULATIVE_ACTIVE_POWER_ID	5805
+#define LIGHT_POWER_FACTOR_ID			5820
+#define LIGHT_COLOUR_ID				5706
+#define LIGHT_SENSOR_UNITS_ID			5701
+
+#define LIGHT_MAX_ID		7
+
+#define MAX_INSTANCE_COUNT	CONFIG_LWM2M_IPSO_LIGHT_CONTROL_INSTANCE_COUNT
+
+#define LIGHT_STRING_SHORT	8
+
+/* resource state variables */
+static bool on_off_value[MAX_INSTANCE_COUNT];
+static u8_t dimmer_value[MAX_INSTANCE_COUNT];
+static s32_t on_time_value[MAX_INSTANCE_COUNT];
+static u32_t on_time_offset[MAX_INSTANCE_COUNT];
+static float32_value_t cumulative_active_value[MAX_INSTANCE_COUNT];
+static float32_value_t power_factor_value[MAX_INSTANCE_COUNT];
+static char colour[MAX_INSTANCE_COUNT][LIGHT_STRING_SHORT];
+static char units[MAX_INSTANCE_COUNT][LIGHT_STRING_SHORT];
+
+static struct lwm2m_engine_obj light_control;
+static struct lwm2m_engine_obj_field fields[] = {
+	OBJ_FIELD_DATA(LIGHT_ON_OFF_ID, RW, BOOL),
+	OBJ_FIELD_DATA(LIGHT_DIMMER_ID, RW, U8),
+	OBJ_FIELD_DATA(LIGHT_ON_TIME_ID, RW, S32),
+	OBJ_FIELD_DATA(LIGHT_CUMULATIVE_ACTIVE_POWER_ID, R, FLOAT32),
+	OBJ_FIELD_DATA(LIGHT_POWER_FACTOR_ID, R, FLOAT32),
+	OBJ_FIELD_DATA(LIGHT_COLOUR_ID, RW, STRING),
+	OBJ_FIELD_DATA(LIGHT_SENSOR_UNITS_ID, R, STRING),
+};
+
+static struct lwm2m_engine_obj_inst inst[MAX_INSTANCE_COUNT];
+static struct lwm2m_engine_res_inst res[MAX_INSTANCE_COUNT][LIGHT_MAX_ID];
+
+static void *on_time_read_cb(u16_t obj_inst_id, size_t *data_len)
+{
+	int i;
+
+	for (i = 0; i < MAX_INSTANCE_COUNT; i++) {
+		if (inst[i].obj && inst[i].obj_inst_id == obj_inst_id) {
```

Comment:
I would reverse this test and if it applies put a continue;

Less if/if/if imbrication.